# HW2

Jaee Oh / Tejas Phanse / Mandadi Ayush Reddy

## Part A: DTM

In [1]:
import nltk
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import re
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from bs4 import BeautifulSoup
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk

nltk.download('stopwords')
nltk.download('wordnet') 
print("Libraries loaded successfully.")

Libraries loaded successfully.


[nltk_data] Downloading package stopwords to /home/jaee/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /home/jaee/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


- Text Prerpcessing

In [2]:
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

# List of common words to be removed from full_text_clean column
common_word_list = {'followers', 'instagram', 'facebook', 'like', 'messages', 'support',
                    'sitemap', 'free', 'faq', 'twitter', 'search', 'subscribe', 'youtube', 
                    'feedback', 'linkedin', 'demo', 'menu', 'following', 'unsubscribe', 
                    'snapchat', 'member', 'tiktok', 'whatsapp', 
                    'profile', 'share', 'reddit', 'download', 'settings', 'notifications', 
                    'home', 'qa', 'people'}
stop_words = stop_words.union(common_word_list) # Add common words to NLTK stop words

# Common phrases to remove
phrase_list = ["find out more", "join now","terms & conditions", "terms and conditions","all rights reserved","free trial", 
             "click here", "about us", "contact us","terms of service", "privacy policy", "help center", "our story","our team",
             "read more", "learn more", "get started", "download now", "sign up", "log in"]

def preprocess(text):
    # Remove \n, \r, \t
    text = re.sub(r'[\n\r\t]+', ' ', text)

    # Remove HTML tags like <body>, but first remove script/style content
    soup = BeautifulSoup(text, 'html.parser')
    for tag in soup(['script', 'style', 'noscript']):
        tag.decompose()
    text = soup.get_text(separator=' ')
    
    # Lowercase
    text = text.lower()
    
    # Remove URLs
    text = re.sub(r'(?:https?://|ftp://|www\.)\S+|mailto:\S+|tel:\S+', '', text)
    
    # Remove emails
    text = re.sub(r'\b[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}\b', '', text)
    
    # Remove phrases
    for phrase in phrase_list:
        text = text.replace(phrase, ' ')

    # Remove non-ASCII characters
    text = re.sub(r'[^\x00-\x7F]', ' ', text)
    
    # Keep only letters and spaces
    text = re.sub(r'[^a-z\s]', '', text)

    # Tokenize and filter with stop words and lemmatization
    tokens = text.split()
    tokens = [lemmatizer.lemmatize(t) for t in tokens if len(t) > 2 and t not in stop_words]
    
    return ' '.join(tokens)

In [3]:
# Load documents
documents = pd.read_csv('../data/dataset_with_assignments.csv')
# Apply preprocessing
documents['text_clean'] = documents['full_text'].apply(preprocess)

# Basic CountVectorizer
count_vec = CountVectorizer()
count_matrix = count_vec.fit_transform(documents['text_clean'])  # Fixed

print(f"Matrix shape: {count_matrix.shape}")
print(f"  - {count_matrix.shape[0]} documents")
print(f"  - {count_matrix.shape[1]} unique terms in vocabulary")

# Be careful with large vocabularies - this could print thousands of words
vocab = count_vec.get_feature_names_out()
print(f"\nFirst 20 terms: {vocab[:20]}")
print(f"Last 20 terms: {vocab[-20:]}")

doc_freq = np.asarray((count_matrix > 0).sum(axis=0)).flatten() # Number of documents that have each terms

print(f"Terms appearing in 1 doc: {(doc_freq == 1).sum()}")
print(f"Terms appearing in <5 docs: {(doc_freq < 5).sum()}")
print(f"Terms appearing in >50% docs: {(doc_freq > len(documents) * 0.5).sum()}")
print(f"Terms appearing in >85% docs: {(doc_freq > len(documents) * 0.85).sum()}")

/tmp/ipykernel_20014/1190997405.py:23: XMLParsedAsHTMLWarning: It looks like you're using an HTML parser to parse an XML document.

Assuming this really is an XML document, what you're doing might work, but you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the Python package 'lxml' installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.

If you want or need to use an HTML parser on this document, you can make this warning go away by filtering it. To do that, run this code before calling the BeautifulSoup constructor:

    from bs4 import XMLParsedAsHTMLWarning
    import warnings

    warnings.filterwarnings("ignore", category=XMLParsedAsHTMLWarning)

  soup = BeautifulSoup(text, 'html.parser')


Matrix shape: (6314, 222858)
  - 6314 documents
  - 222858 unique terms in vocabulary

First 20 terms: ['aa' 'aaa' 'aaaa' 'aaaaa' 'aaaaaa' 'aaaaaaa' 'aaaaw' 'aaac' 'aaacaa'
 'aaacfbdburvncadkwicddxdqopvauvlooxgnevivzifwevajfhogfkaolrjuzbptdnxjmewmubhdorickodttxwq'
 'aaacsl' 'aaacsldr' 'aaacslz' 'aaacslza' 'aaacslzb' 'aaacslzc' 'aaacslzd'
 'aaacslze' 'aaacslzg' 'aaacslzi']
Last 20 terms: ['zytek' 'zythophile' 'zythophilic' 'zyto' 'zyw' 'zywave' 'zyx' 'zyxel'
 'zza' 'zzaa' 'zzam' 'zzgl' 'zzi' 'zzkc' 'zzmike' 'zzril' 'zzshocks'
 'zztires' 'zzz' 'zzzzz']
Terms appearing in 1 doc: 156034
Terms appearing in <5 docs: 197062
Terms appearing in >50% docs: 1
Terms appearing in >85% docs: 0


After preprocessing, about 70% of the terms (156,034 out of 222,858) appeared in one document, which are most likely irrelavant words, which even includes typos. These terms were hard to be removed by the preprocessing.

Only one term appeared in more than 50% of the documents, which are expected since the original data is not scraped based on a given topic. This also shows that the preprocessing has successfully removed sementically less meaningful words such as a, the, this, that, etc.

This can be our guideline to set our max_df and min_df.

- TF-IDF Vectorizer

We start by setting max_df=0.5 and min_df=5 found from above. max_df=0.5 only removes 1 term, but lowering max_df below that may remove some semantically meaningful terms, so we will stay at max_df=0.5

In [13]:
tfidf_vec = TfidfVectorizer(
    max_df=0.5,      # Ignore terms in >85%
    min_df=5,         # Ignore terms in <5 docs
    # ngram_range=(1,2), # Unigrams and bigrams
    # max_features=10000  # Max vocabulary size
)
tfidf_matrix = tfidf_vec.fit_transform(documents['text_clean'])

print(f"TF-IDF Matrix shape: {tfidf_matrix.shape}")
print(f"Average non-zeros per document: {tfidf_matrix.nnz/tfidf_matrix.shape[0]}")

density = tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1])
print(f"Sparsity: {(1-density) * 100:.4f}%")

TF-IDF Matrix shape: (6314, 25795)
Average non-zeros per document: 224.09550205891668
Sparsity: 99.1312%


In [14]:
tfidf_vec = TfidfVectorizer(
    max_df=0.5,      # Ignore terms in >85%
    min_df=5,         # Ignore terms in <5 docs
    ngram_range=(1,2), # Unigrams and bigrams
    # max_features=10000  # Max vocabulary size
)
tfidf_matrix = tfidf_vec.fit_transform(documents['text_clean'])

print(f"TF-IDF Matrix shape: {tfidf_matrix.shape}")
print(f"Average non-zeros per document: {tfidf_matrix.nnz/tfidf_matrix.shape[0]}")

density = tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1])
print(f"Sparsity: {(1-density) * 100:.4f}%")

TF-IDF Matrix shape: (6314, 75469)
Average non-zeros per document: 308.9257206208426
Sparsity: 99.5907%


In [18]:
tfidf_vec = TfidfVectorizer(
    max_df=0.5,      # Ignore terms in >85%
    min_df=5,         # Ignore terms in <5 docs
    ngram_range=(1,2), # Unigrams and bigrams
    max_features=10000  # Max vocabulary size
)
tfidf_matrix = tfidf_vec.fit_transform(documents['text_clean'])

print(f"TF-IDF Matrix shape: {tfidf_matrix.shape}")
print(f"Average non-zeros per document: {tfidf_matrix.nnz/tfidf_matrix.shape[0]}")

density = tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1])
print(f"Sparsity: {(1-density) * 100:.4f}%")


vocab = tfidf_vec.get_feature_names_out() # Term names
doc_freq = np.asarray((tfidf_matrix > 0).sum(axis=0)).flatten() # Number of documents that have each terms

n = 30
top_idx = doc_freq.argsort()[::-1][:n]

print("Most frequent terms (by document count):")
for i, idx in enumerate(top_idx, 1):
    print(f"  {i}. {vocab[idx]}: {doc_freq[idx]} docs")

TF-IDF Matrix shape: (6314, 10000)
Average non-zeros per document: 211.3297434273044
Sparsity: 97.8867%
Most frequent terms (by document count):
  1. contact: 2902 docs
  2. use: 2819 docs
  3. email: 2729 docs
  4. content: 2715 docs
  5. service: 2629 docs
  6. time: 2549 docs
  7. get: 2440 docs
  8. information: 2439 docs
  9. help: 2259 docs
  10. may: 2231 docs
  11. one: 2144 docs
  12. copyright: 2123 docs
  13. page: 2096 docs
  14. news: 2092 docs
  15. view: 2058 docs
  16. top: 2047 docs
  17. website: 2009 docs
  18. product: 1969 docs
  19. skip: 1964 docs
  20. year: 1960 docs
  21. also: 1956 docs
  22. site: 1905 docs
  23. make: 1884 docs
  24. day: 1872 docs
  25. work: 1872 docs
  26. first: 1870 docs
  27. link: 1862 docs
  28. need: 1823 docs
  29. please: 1817 docs
  30. blog: 1806 docs


By adding bigrams, our density of non-zero terms were lowered by twice. Then, we picked only 10,000 max terms of the matrix, and the density is now 2%, which is 5 times increased. This shows that reducing max_feature from 75,469 to 10,000 mostly reduced zero terms.